service/spi/src/test/java/org/kaazing/gateway/service/messaging/collections/MemoryTopicTest.java

```diff
@@ -0,0 +1,175 @@
+        listenersCalled.await();
+        assertEquals(1, topic.getLocalTopicStats().getPublishOperationCount());
+        assertEquals(3, topic.getLocalTopicStats().getReceiveOperationCount());
+    }
+
+    @Test
+    public void shouldCallMultipleTimesMessageListener() throws InterruptedException {
+        ITopic<String> topic = factory.getTopic("topic");
+        CountDownLatch listenersCalled = new CountDownLatch(2);
+        MessageListener m = message -> listenersCalled.countDown();
+        topic.addMessageListener(m);
+        topic.addMessageListener(m);
+        topic.publish("msg1");
+        listenersCalled.await();
+        assertEquals(1, topic.getLocalTopicStats().getPublishOperationCount());
+        assertEquals(2, topic.getLocalTopicStats().getReceiveOperationCount());
+    }
+
+    @Test
+    public void shouldAddAndRemoveMessageListener() throws InterruptedException {
+        ITopic<String> topic = factory.getTopic("topic");
+        CountDownLatch listenersCalled = new CountDownLatch(1);
+        MessageListener m = message -> listenersCalled.countDown();
+        String name = topic.addMessageListener(m);
+        topic.publish("msg1");
+        listenersCalled.await();
+        topic.removeMessageListener(name);
+        topic.publish("msg2");
+        assertEquals(2, topic.getLocalTopicStats().getPublishOperationCount());
+        assertEquals(1, topic.getLocalTopicStats().getReceiveOperationCount());
+    }
+
+
+    @Test
+    public void shouldAddAndRemoveMessageListenerWithoutDeadlock() throws InterruptedException {
+        ITopic<String> topic = factory.getTopic("topic");
+        CountDownLatch listenersCalled = new CountDownLatch(1);
+        MessageListener m1 = message -> {
+            listenersCalled.countDown();
+        };
+        String name = topic.addMessageListener(m1);
+        MessageListener m2 = message -> {
+            topic.removeMessageListener(name);
+        };
+        topic.addMessageListener(m2);
+        topic.publish("msg1");
+        listenersCalled.await();
+    }
+
+
+    @Test
+    public void shouldAddAndRemoveMessageListenerNoSideEffects() throws InterruptedException {
+        ITopic<String> topic = factory.getTopic("topic");
+        CountDownLatch listenersCalled = new CountDownLatch(2);
+        AtomicBoolean calledM3 = new AtomicBoolean(false);
+        MessageListener m1 = message -> {
+            listenersCalled.countDown();
+            topic.addMessageListener(message1 -> calledM3.set(true));
+        };
+        String name = topic.addMessageListener(m1);
+        MessageListener m2 = message -> {
+            topic.removeMessageListener(name);
+            listenersCalled.countDown();
+        };
+        topic.addMessageListener(m2);
+        topic.publish("msg1");
+        listenersCalled.await();
+        assertTrue("MessageListener m3 must not be called", calledM3.get() == false);
+    }
+
+    @Test
+    public void shouldPubSubFromMessageListeners() {
```

Comment:
We won't actually need to do this so we don't need to support it. Would be interesting to know if Hazelast supports it. We'll find out when you do the Hazelcast variant of the test.